capsules/src/rf233.rs

```diff
@@ -0,0 +1,882 @@
+            self.tx_buf.replace(_write);
+            self.state.set(InternalState::TX_WRITING_FRAME_DONE);
+        } else {
+            self.spi_rx.replace(rbuf);
+            self.spi_tx.replace(_write);
+        }
+
+        // This first case is when an interrupt fired during an SPI operation:
+        // we wait for the SPI operation to complete then handle the
+        // interrupt by reading the IRQ_STATUS register over the SPI.
+        // Since itself is an SPI operation, return.
+        if self.interrupt_pending.get() == true {
+            self.interrupt_pending.set(false);
+            self.handle_interrupt();
+            return;
+        }
+        // This second case is when the SPI operation is reading the
+        // IRQ_STATUS register from handling an interrupt. Note that
+        // we're done handling the interrupt and continue with the
+        // state machine. This is an else because handle_interrupt
+        // sets interrupt_handling to true.
+        if handling {
+            self.interrupt_handling.set(false);
+            let state = self.state.get();
+            let interrupt = result;
+            if state == InternalState::ON_PLL_WAITING {
+                if interrupt_included(interrupt, IRQ_0_PLL_LOCK) {
+                    self.state.set(InternalState::ON_PLL_SET);
+                }
+            } else if (state == InternalState::TX_TRANSMITTING) &&
+                      interrupt_included(interrupt, IRQ_3_TRX_END) {
+                self.state.set(InternalState::TX_DONE);
+            }
+            if interrupt_included(interrupt, IRQ_2_RX_START) {
+                // Start of frame
+                self.receiving.set(true);
+                self.state.set(InternalState::RX);
+            }
+
+            // We've received  an entire frame into the frame buffer.
+            // There are three cases:
+            //   1. we have a receive buffer: copy it out
+            //   2. no receive buffer, but transmission pending: send
+            //   3. no receive buffer, no transmission: return to waiting
+            if (interrupt_included(interrupt, IRQ_3_TRX_END) && self.receiving.get()) {
+                self.receiving.set(false);
+                if self.rx_buf.is_some() {
+                    self.state.set(InternalState::RX_START_READING);
+                } else if self.transmitting.get() {
+                    self.state_transition_read(RF233Register::MIN,
+                                               InternalState::TX_STATUS_PRECHECK1);
+                    return;
+                } else {
+                    self.state_transition_read(RF233Register::TRX_STATUS, InternalState::READY);
+                    return;
+                }
+            }
+
+        }
+
+        match self.state.get() {
+            // Default on state; wait for transmit() call or receive
+            // interrupt
+            InternalState::READY => {
+                self.radio_on.set(true);
```

Comment:
Yes -- although app_buf is no longer declared. But that is not hard to add if someone wants to use this code.